File: classes/circle.js
import Force from "./force.js";

class Circle {
  constructor(particles) {
    this.particles = particles;
    this.x = 0;
    this.y = 0;
    this.r = 1;
  }
  setPositionAndRadius(x, y, r) {
    this.x = x;
    this.y = y;
    this.r = r;
  }
  getParticles() {
    const cy = this.y;
    const cx = this.x;
    const r = this.r;

    const particles_inside = this.particles.filter((particle) => {
      const x = particle.position.x;
      const y = particle.position.y;
      const inside = (cx - x) ** 2 + (cy - y) ** 2 < r ** 2;
      return inside;
    });

    const particles_updateable = particles_inside.filter((particle) => {
      return !particle.isUpdating;
    });
    return { inside: particles_inside, updateable: particles_updateable };
  }
  update(updateWrapper, callbackFunc) {
    const { inside, updateable } = this.getParticles();

    let newUpdateWrapper;
    if (typeof updateWrapper === "function") {
      newUpdateWrapper = (runUpdate) => {
        updateWrapper(
          runUpdate,
          (updateable.length) **
            Force.inefficienctyFactor / (Force.inefficienctyFactor ** 2)
        );
      };
    }
    const forces = updateable.map((p) => {
      return p.getForce(this.particles);
    });
    const runUpdate = () => {
      for (let i in updateable) {
        const particle = updateable[i];
        const force = forces[i];
        particle.update(force, newUpdateWrapper);
      }
      callbackFunc();
    };
    runUpdate();
  }
}

export default Circle;

File: threaded/circle.py
import time
import threading

class Circle:
  def __init__(self, particles):
    self.particles = particles
    pass

  def setPositionAndRadius(self, x, y, r):
    self.x = x
    self.y = y
    self.r = r
  
  def getParticles(self):
    cy = self.y
    cx = self.x
    r = self.r

    particles_inside = [
      particle for particle in self.particles
      if (cx - particle.position['x']) ** 2 + (cy - particle.position['y']) ** 2 < r ** 2
    ]

    particles_updateable = []
    for particle in particles_inside:
      if particle.acquire(False):
        particles_updateable.append(particle)

    return particles_updateable
    return {'inside': particles_inside, 'updateable': particles_updateable}

  def update(self):
    particles = self.getParticles()
    forces = [p.getForce(self.particles) for p in particles]

    for i, particle in enumerate(particles):
      force = forces[i]
      particle.update(force)
      particle.release()
    
    return len(particles)